src/autocomplete/zsh.ts

```diff
@@ -0,0 +1,400 @@
+    // foobar -x -- -y
+    // the ‘-x’ is considered a flag, the ‘-y’ is considered an argument, and the ‘--’ is considered to be neither.
+    let argumentsBlock = '_arguments -S \\\n'
+
+    for (const flagName of flagNames) {
+      const f = flags[flagName]
+
+      // skip hidden flags
+      if (f.hidden) continue
+
+      f.summary = sanitizeSummary(f.summary || f.description)
+
+      let flagSpec = ''
+
+      if (f.type === 'option') {
+        if (f.char) {
+          if (f.multiple) {
+            // this flag can be present multiple times on the line
+            flagSpec += `"*"{-${f.char},--${f.name}}`
+          } else {
+            flagSpec += `"(-${f.char} --${f.name})"{-${f.char},--${f.name}}`
+          }
+
+          flagSpec += `"[${f.summary}]`
+
+          if (f.options) {
+            flagSpec += `:${f.name} options:(${f.options?.join(' ')})"`
+          } else {
+            flagSpec += ':file:_files"'
+          }
+        } else {
+          if (f.multiple) {
+            // this flag can be present multiple times on the line
+            flagSpec += '"*"'
+          }
+
+          flagSpec += `--${f.name}"[${f.summary}]:`
+
+          if (f.options) {
+            flagSpec += `${f.name} options:(${f.options.join(' ')})"`
+          } else {
+            flagSpec += 'file:_files"'
+          }
+        }
+      } else if (f.char) {
+        // Flag.Boolean
+        flagSpec += `"(-${f.char} --${f.name})"{-${f.char},--${f.name}}"[${f.summary}]"`
+      } else {
+        // Flag.Boolean
+        flagSpec += `--${f.name}"[${f.summary}]"`
+      }
+
+      flagSpec += ' \\\n'
+      argumentsBlock += flagSpec
+    }
+    // complete files if `-` is not present on the current line
+    argumentsBlock += '"*: :_files"'
+
+    return argumentsBlock
+  }
+
+  private genZshValuesBlock(subArgs: {id: string; summary?: string}[]): string {
+    let valuesBlock = '_values "completions" \\\n'
+
+    subArgs.forEach(subArg => {
+      valuesBlock += `"${subArg.id}[${subArg.summary}]" \\\n`
+    })
+
+    return valuesBlock
+  }
+
+  private genZshTopicCompFun(id: string): string {
+    const coTopics: string[] = []
+
+    for (const topic of this.topics) {
+      for (const cmd of this.commands) {
+        if (topic.name === cmd.id) {
+          coTopics.push(topic.name)
+        }
+      }
+    }
+
+    const flagArgsTemplate = '        "%s")\n          %s\n        ;;\n'
+
+    const underscoreSepId = id.replace(/:/g, '_')
+    const depth = id.split(':').length
+
+    const isCotopic = coTopics.includes(id)
+
+    if (isCotopic) {
+      const compFuncName = `${this.config.bin}_${underscoreSepId}`
+
+      const coTopicCompFunc =
+`_${compFuncName}() {
+  _${compFuncName}_flags() {
+    local context state state_descr line
+    typeset -A opt_args
+
+    ${this.genZshFlagArgumentsBlock(this.commands.find(c => c.id === id)?.flags)}
+  }
+
+  local context state state_descr line
```

Comment:
can't link to zsh docs so I'll leave chunks related to the code that are searchable here:
https://zsh.sourceforge.io/Doc/Release/Completion-System.html

`A function calling _arguments with at least one action containing a ‘->string’ must therefore declare appropriate local parameters:`